### DISCUSSION TITLE
rust clang build

### DISCUSSION THREAD

**Discussion by revelator** [msys2/MINGW-packages#9439] [Q&A]
any headway on this one ?.

**Comment by mati865** [comment]
There is mingw-w64-clang-x86_64-rust, 32 bit is crashing during build.

**Reply by revelator** [reply]
ah ok.

**Reply by jeremyd2019** [reply]
#9092 - managed to get psm build fixed, but then generated binaries crash.

**Reply by jeremyd2019** [reply]
I came back upon this today, as my script to enable clang32 on packages is getting diminishing returns, and investigating a bit found gtk3 depends on adwaita-icon-theme which depends on librsvg (and gtk3 but presumably that's easily resolved).  If rust is going to be intractable maybe need to get the C version of librsvg packaged to unblock?

**Reply by revelator** [reply]
hmm interresting i use the older c based librsvg and so far rust seems to build now, it is slow going though but as far as i heard that is pretty normal with rust.

**Reply by revelator** [reply]
ah no got further than ever but still crashes :/

```
   Compiling handlebars v3.4.0
error: failed to run custom build command for `html5ever v0.25.1`

Caused by:
  process didn't exit successfully: `I:\mingw-w64-rust\src\i686\build\i686-pc-windows-gnu\stage0-bootstrap-tools\release\build\html5ever-7b859b6107fbb733\build-script-build` (exit code: 0xc0000409, STATUS_STACK_BUFFER_OVERRUN)
  --- stdout
  cargo:rerun-if-changed=C:\Users\ralph\.cargo\registry\src\github.com-1285ae84e5963aae\html5ever-0.25.1\src/tree_builder/rules.rs

  --- stderr
  fatal runtime error: failed to initiate panic, error 5
warning: build failed, waiting for other jobs to finish...
error: build failed
command did not execute successfully: "\\\\?\\I:\\mingw-w64-rust\\src\\i686\\build\\i686-pc-windows-gnu\\stage0\\bin\\cargo.exe" "build" "--target" "i686-pc-windows-gnu" "-Zbinary-dep-depinfo" "-j" "12" "--release" "--manifest-path" "I:\\mingw-w64-rust\\src\\rustc-1.53.0-src\\src/tools/rustbook\\Cargo.toml" "--message-format" "json-render-diagnostics"
expected success, got: exit code: 101
Traceback (most recent call last):
  File "I:\mingw-w64-rust\src\rustc-1.53.0-src\x.py", line 27, in <module>
    bootstrap.main()
  File "I:/mingw-w64-rust/src/rustc-1.53.0-src/src/bootstrap/bootstrap.py", line 1189, in main
    bootstrap(help_triggered)
  File "I:/mingw-w64-rust/src/rustc-1.53.0-src/src/bootstrap/bootstrap.py", line 1175, in bootstrap
    run(args, env=env, verbose=build.verbose)
  File "I:/mingw-w64-rust/src/rustc-1.53.0-src/src/bootstrap/bootstrap.py", line 145, in run
    raise RuntimeError(err)
RuntimeError: failed to run: I:/mingw-w64-rust/src/i686/build/bootstr
ap/debug/bootstrap install --stage 2 --verbose
```

**Reply by jeremyd2019** [reply]
yep, that's what I get too.

**Reply by jeremyd2019** [reply]
I wonder if https://github.com/thepowersgang/mrustc would have better luck.  At least it's written in C++ and generates C as output, which are both languages I know (as opposed to Rust).

**Reply by revelator** [reply]
possibly though depends on if it is even builable with our clang toolchain i reckon ?

**Reply by revelator** [reply]
might be simply some incompatibility with the ucrt runtime, as far as i know we are the only ones attempting it.
maybe a good idea to post the rust devs or ask them to come by here and take a look at things.

**Reply by revelator** [reply]
hmm error 5 is Input/output error according to the devs.

**Reply by mati865** [reply]
I doubt it's IO, this comment seems relevant https://github.com/servo/html5ever/blob/0e03e1c2b1f63e81f831fd95b9eb8bbde18b7815/html5ever/build.rs#L24
I was able to hack together i686 clang build but it was crashing when running proc macros.
Proc macros parse and alter source code so stack size issue sounds likely except html5ever already tries 128MiB (not sure though if setting stack size on Windows does work).

**Reply by revelator** [reply]
could be a stack problem, not sure if setting anything via msys2's shell will affect the windows clang compiler atleast it does nothing for mingw gcc so else we might have to do it codewise.

**Reply by jeremyd2019** [reply]
My entirely unsupported by evidence concern (other than the fact that we hit an error building psm that apparently nobody else has) is that maybe it's exercising a different mechanism in psm/stacker that doesn't work properly.  IIRC it *should* be using fibers on Windows.

**Reply by revelator** [reply]
would it be possible to disable just this module to see if the rust compiler itself works ?, i cant say why but i have a hunch it might actually.

**Reply by jeremyd2019** [reply]
I tried disabling building docs, and it failed in something else.  I think @mati865 said he disabled other things and got a compiler to build, but the code it generated was broken.

**Reply by revelator** [reply]
ah ok, damn :S sadly i have no idea about rust code just playing of a hunch.

**Reply by jeremyd2019** [reply]
Me neither, which makes this whole thing all the more frustrating, I can get a stack trace (https://github.com/msys2/MINGW-packages/pull/9092#issuecomment-878678348) but have no idea how to actually debug rust code.

**Reply by revelator** [reply]
aye me either :(, might be just a slight oversight might be something more profound, i think we might need the help of the rust devs to get this one going.

**Reply by jeremyd2019** [reply]
For fun I tried updating to rust 1.54.0 but same problem.

**Reply by revelator** [reply]
kinda odd though that a compiler which relies heavily on the llvm toolchain needs gcc to actually build xD

**Reply by revelator** [reply]
off for a few days gotta visit the my family :) ill try getting them on the vagon when i return.

**Reply by mati865** [reply]
I'll try to return to it on the weekend when I'll be back home.

**Reply by raedrizqie** [reply]
maybe reverting this [commit](https://github.com/rust-lang/rust/pull/37373/commits/0a16a11c392d227ca845e65905c06599579e2828) will fix clang32 build?

**Reply by mati865** [reply]
I think it's entirely unrelated, the problem is exceptions handling is broken.
